net/src/tcp/truncated.rs

```diff
@@ -0,0 +1,185 @@
+// Copyright Open Network Fabric Authors
+
+//! TCP header type and logic, for potentially truncated datagrams.
+
+use std::num::NonZero;
+
+use crate::parse::{DeParse, DeParseError, LengthError, Parse, ParseError};
+use crate::tcp::{Tcp, TcpParseError, TcpPort};
+
+/// A truncated TCP header.
+///
+/// This truncated header is built from the start of a regular TCP header, down to the last byte of
+/// the packet, but does not contain a full header. The only fields that are guaranteed to be
+/// present are the source and destination ports.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct TruncatedTcpHeader {
+    source_port: TcpPort,
+    destination_port: TcpPort,
+    // The rest of the header, as a byte vector, for de-parsing
+    everything_else: Vec<u8>,
+}
+
+impl TruncatedTcpHeader {
+    const MIN_HEADER_LEN: usize = 4;
+
+    fn new(source_port: TcpPort, destination_port: TcpPort, everything_else: Vec<u8>) -> Self {
+        Self {
+            source_port,
+            destination_port,
+            everything_else,
+        }
+    }
+
+    /// Get the length of the truncated header
+    #[must_use]
+    pub fn header_len(&self) -> NonZero<usize> {
+        let len = self.everything_else.len() + Self::MIN_HEADER_LEN;
+        NonZero::new(len).unwrap_or_else(|| unreachable!())
+    }
+
+    /// Get the source port
+    #[must_use]
+    pub const fn source(&self) -> TcpPort {
+        self.source_port
+    }
+
+    /// Get the destination port
+    #[must_use]
+    pub const fn destination(&self) -> TcpPort {
+        self.destination_port
+    }
+
+    /// Set the source port
+    pub fn set_source(&mut self, source_port: TcpPort) -> &mut Self {
+        self.source_port = source_port;
+        self
+    }
+
+    /// Set the destination port
+    pub fn set_destination(&mut self, destination_port: TcpPort) -> &mut Self {
+        self.destination_port = destination_port;
+        self
+    }
+}
+
+impl Parse for TruncatedTcpHeader {
+    type Error = TruncatedTcpError;
+
+    fn parse(buf: &[u8]) -> Result<(Self, NonZero<u16>), ParseError<Self::Error>> {
+        // We need at least four bytes to form our truncated header.
+        // RFC 792 (ICMP) says embedded packets in ICMP Error messages contain the IP header plus at
+        // least the first 64 bits from the datagram, so we should have these 4 bytes. Otherwise,
+        // it's an error.
+        if buf.len() < TruncatedTcpHeader::MIN_HEADER_LEN {
+            return Err(ParseError::Length(LengthError {
+                expected: NonZero::new(TruncatedTcpHeader::MIN_HEADER_LEN)
+                    .unwrap_or_else(|| unreachable!()),
+                actual: buf.len(),
+            }));
+        }
+
+        let parsed_source_port = u16::from_be_bytes([buf[0], buf[1]]);
+        let parsed_destination_port = u16::from_be_bytes([buf[2], buf[3]]);
+        let consumed = NonZero::new(u16::try_from(buf.len()).unwrap_or_else(|_| unreachable!()))
```

Comment:
Why not
```
NonZero::new(buf.len() as u16).unwrap()
```
since we really don't expect this to panic and suppress the clippy warning about unwrap?
This is simpler and we already assume that buffers cannot exceed 64k, right? This strip the upper order bits of buffer of course, but that is probably fine here.